test/metabase/test/data/drill.clj

```diff
@@ -0,0 +1,177 @@
+           java.sql.SQLException))
+
+(def ^:const field-base-type->sql-type
+  {:type/BigInteger "BIGINT"
+   :type/Boolean    "BOOLEAN"
+   :type/Date       "DATE"
+   :type/DateTime   "TIMESTAMP"
+   :type/Decimal    "DECIMAL"
+   :type/Float      "DOUBLE"
+   :type/Integer    "INTEGER"
+   :type/Text       "VARCHAR"})
+
+(defn- dash-to-underscore [s]
+  (s/replace s #"-" "_"))
+
+(defn- qualified-name-components
+  ([db-name]                       (map dash-to-underscore [db-name]))
+  ([db-name table-name]            (map dash-to-underscore [db-name table-name]))
+  ([db-name table-name field-name] (map dash-to-underscore [table-name field-name])))
+
+(defn database->connection-details [context {:keys [database-name]}]
+  (merge {:drill-connect "drillbit=localhost;schema=dfs.tmp"}))
+
+(defn quote-name [nm]
+  (str \` nm \`))
+
+(defn make-sure-file-exists! [path]
+  (let [file-path (clojure.java.io/file path)]
+    (when-not (.exists file-path)
+      (.createNewFile file-path))))
+
+(defn create-table-sql [{:keys [database-name], :as dbdef} {:keys [table-name field-definitions]}]
+  (let [qualified-table-name (i/db-qualified-table-name database-name table-name)
+        table-path (str "/tmp/" qualified-table-name "_table.csv")]
+    ;; we can't create the view unless the file exists, so create an empty one if necessary
+    (make-sure-file-exists! table-path)
+    (format "CREATE OR REPLACE VIEW %s AS SELECT %s FROM dfs.`%s`"
+            (str "dfs.tmp.`" qualified-table-name "`")
+            (str (->> field-definitions
+                      (map (fn [{:keys [field-name base-type]}]
+                             (if (= base-type :type/DateTime)
+                               (format "CAST(to_timestamp(`%s`, 'YYYY-MM-dd''T''HH:mm:ss.SSSZ') AS TIMESTAMP) AS `%s`"
+                                       field-name field-name)
+                               (format "CAST(`%s` AS %s) AS `%s`"
+                                       field-name
+                                       (field-base-type->sql-type base-type)
+                                       field-name))))
+                      (interpose ", ")
+                      (apply str))
+                 ", CAST(`id` as INTEGER) AS `id` ")
+            table-path)))
+
+(defn drop-table-if-exists-sql [{:keys [database-name]} {:keys [table-name]}]
+  (format "DROP VIEW IF EXISTS %s" (str "dfs.tmp.`" database-name "_" table-name "`")))
+
+;; workaround for DRILL-5136 (can't use prepared statements for DDL)
+;; instead of prepared statements, we use plain statements.
+(defn sequentially-execute-sql!
+  "Alternative implementation of `execute-sql!` that executes statements one at a time for drivers
+   that don't support executing multiple statements at once.
+
+   Since there are some cases were you might want to execute compound statements without splitting, an upside-down ampersand (`⅋`) is understood as an
+   \"escaped\" semicolon in the resulting SQL statement."
```

Comment:
Also at any rate worth repeating but if you copy-paste a function and change what it does you need to change the docstring too. Docstrings that are out-of-date or wrong are worse than no docstrings at all. Since other people might be looking at this driver as an example of how they write their own we want to make their lives a bit easier and make sure the code is clean and readable.

https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-sure-to-update-comments

Reply:
Done